File: DevExpress-Demo/DevExpress.HybridApp.Wpf/CS/DevExpress.HybridApp.Wpf/Filtering/ViewModel/FiltersSettings.cs
using DevExpress.DevAV.Common.ViewModel;
using DevExpress.DevAV.DevAVDbDataModel;
using DevExpress.DevAV.Properties;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.ViewModel;
using System;

namespace DevExpress.DevAV.ViewModels {
    internal static class FiltersSettings {
        public static FilterTreeViewModel<Employee, long> GetDashboardFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Employee, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, null, null, null, null),
                CreateUnitOfWork().Employees, new Action<object, Action>(RegisterEntityChangedMessageHandler<Employee, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<EmployeeTask, long> GetTasksFilterTree(object parentViewModel) {
            return FilterTreeViewModel<EmployeeTask, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, null, x => x.TasksStaticFilters, null, null),
                CreateUnitOfWork().Tasks, new Action<object, Action>(RegisterEntityChangedMessageHandler<EmployeeTask, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<Employee, long> GetEmployeesFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Employee, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, "Status", x => x.EmployeesStaticFilters, null, null),
                CreateUnitOfWork().Employees, new Action<object, Action>(RegisterEntityChangedMessageHandler<Employee, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<Product, long> GetProductsFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Product, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, "Category", x => x.ProductsStaticFilters, x => x.ProductsCustomFilters,
                    new[] {
                        BindableBase.GetPropertyName(() => new Product().Engineer),
                        BindableBase.GetPropertyName(() => new Product().Brochure),
                        BindableBase.GetPropertyName(() => new Product().Support),
                        BindableBase.GetPropertyName(() => new Product().PrimaryImage),
                        BindableBase.GetPropertyName(() => new Product().ProductImage),
                    }),
                CreateUnitOfWork().Products, new Action<object, Action>(RegisterEntityChangedMessageHandler<Product, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<Customer, long> GetCustomersFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Customer, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, "Favorites", null, x => x.CustomersCustomFilters,
                    new[] {
                        BindableBase.GetPropertyName(() => new Customer().Id),
                        BindableBase.GetPropertyName(() => new Customer().BillingAddress),
                        BindableBase.GetPropertyName(() => new Customer().HomeOffice),
                        BindableBase.GetPropertyName(() => new Customer().Image),
                        BindableBase.GetPropertyName(() => new Customer().Website),
                    },
                    new[] {
                        BindableBase.GetPropertyName(() => new Customer().BillingAddress) + "." + BindableBase.GetPropertyName(() => new Address().City),
                        BindableBase.GetPropertyName(() => new Customer().BillingAddress) + "." + BindableBase.GetPropertyName(() => new Address().State),
                        BindableBase.GetPropertyName(() => new Customer().BillingAddress) + "." + BindableBase.GetPropertyName(() => new Address().ZipCode),
                    }),
                CreateUnitOfWork().Customers, new Action<object, Action>(RegisterEntityChangedMessageHandler<Customer, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<Order, long> GetSalesFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Order, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, null, null, null, null),
                CreateUnitOfWork().Orders.ActualOrders(), new Action<object, Action>(RegisterEntityChangedMessageHandler<Order, long>)
            ).SetParentViewModel(parentViewModel);
        }
        public static FilterTreeViewModel<Quote, long> GetOpportunitiesFilterTree(object parentViewModel) {
            return FilterTreeViewModel<Quote, long>.Create(
                new FilterTreeModelPageSpecificSettings<Settings>(Settings.Default, null, null, null, null),
                CreateUnitOfWork().Quotes.ActualQuotes(), new Action<object, Action>(RegisterEntityChangedMessageHandler<Quote, long>)
            ).SetParentViewModel(parentViewModel);
        }

        static IDevAVDbUnitOfWork CreateUnitOfWork() {
            return UnitOfWorkSource.GetUnitOfWorkFactory().CreateUnitOfWork();
        }

        static void RegisterEntityChangedMessageHandler<TEntity, TPrimaryKey>(object recipient, Action handler) {
            Messenger.Default.Register<EntityMessage<TEntity, TPrimaryKey>>(recipient, message => handler());
        }
    }
}

File: DevExpress-Demo/DevExpress.HybridApp.Wpf/VB/DevExpress.HybridApp.Wpf/Filtering/ViewModel/FiltersSettings.vb
Imports DevExpress.DevAV.DevAVDbDataModel
Imports DevExpress.DevAV.Properties
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.ViewModel
Imports System

Namespace DevExpress.DevAV.ViewModels

    Friend Module FiltersSettings

        Public Function GetDashboardFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Employee, Long)
            Return FilterTreeViewModel(Of Employee, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, Nothing, Nothing, Nothing, Nothing), CreateUnitOfWork().Employees, New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Employee, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetTasksFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of EmployeeTask, Long)
            Return FilterTreeViewModel(Of EmployeeTask, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, Nothing, Function(x) x.TasksStaticFilters, Nothing, Nothing), CreateUnitOfWork().Tasks, New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of EmployeeTask, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetEmployeesFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Employee, Long)
            Return FilterTreeViewModel(Of Employee, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, "Status", Function(x) x.EmployeesStaticFilters, Nothing, Nothing), CreateUnitOfWork().Employees, New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Employee, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetProductsFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Product, Long)
            Return FilterTreeViewModel(Of Product, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, "Category", Function(x) x.ProductsStaticFilters, Function(x) x.ProductsCustomFilters, {BindableBase.GetPropertyName(Function() New Product().Engineer), BindableBase.GetPropertyName(Function() New Product().Brochure), BindableBase.GetPropertyName(Function() New Product().Support), BindableBase.GetPropertyName(Function() New Product().PrimaryImage), BindableBase.GetPropertyName(Function() New Product().ProductImage)}), CreateUnitOfWork().Products, New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Product, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetCustomersFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Customer, Long)
            Return FilterTreeViewModel(Of Customer, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, "Favorites", Nothing, Function(x) x.CustomersCustomFilters, {BindableBase.GetPropertyName(Function() New Customer().Id), BindableBase.GetPropertyName(Function() New Customer().BillingAddress), BindableBase.GetPropertyName(Function() New Customer().HomeOffice), BindableBase.GetPropertyName(Function() New Customer().Image), BindableBase.GetPropertyName(Function() New Customer().Website)}, {BindableBase.GetPropertyName(Function() New Customer().BillingAddress) & "." & BindableBase.GetPropertyName(Function() New Address().City), BindableBase.GetPropertyName(Function() New Customer().BillingAddress) & "." & BindableBase.GetPropertyName(Function() New Address().State), BindableBase.GetPropertyName(Function() New Customer().BillingAddress) & "." & BindableBase.GetPropertyName(Function() New Address().ZipCode)}), CreateUnitOfWork().Customers, New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Customer, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetSalesFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Order, Long)
            Return FilterTreeViewModel(Of Order, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, Nothing, Nothing, Nothing, Nothing), CreateUnitOfWork().Orders.ActualOrders(), New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Order, Long))).SetParentViewModel(parentViewModel)
        End Function

        Public Function GetOpportunitiesFilterTree(ByVal parentViewModel As Object) As FilterTreeViewModel(Of Quote, Long)
            Return FilterTreeViewModel(Of Quote, Long).Create(New FilterTreeModelPageSpecificSettings(Of Settings)(Settings.Default, Nothing, Nothing, Nothing, Nothing), CreateUnitOfWork().Quotes.ActualQuotes(), New Action(Of Object, Action)(AddressOf RegisterEntityChangedMessageHandler(Of Quote, Long))).SetParentViewModel(parentViewModel)
        End Function

        Private Function CreateUnitOfWork() As IDevAVDbUnitOfWork
            Return GetUnitOfWorkFactory().CreateUnitOfWork()
        End Function

        Private Sub RegisterEntityChangedMessageHandler(Of TEntity, TPrimaryKey)(ByVal recipient As Object, ByVal handler As Action)
            Call Messenger.Default.Register(Of EntityMessage(Of TEntity, TPrimaryKey))(recipient, Sub(message) handler())
        End Sub
    End Module
End Namespace
